[expo-font][android] Rewrite android code to Kotlin

kkafar opened the pull request:
# Why

This task was done as a part of migrating expo modules to Kotlin project.

# How

The code had been initially converted from Java to Kotlin with help of Android Studio's conversion tool. Then I was manually adjusted / improved.

# Test Plan

BareExpo (test-suite + ncl)

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).

kkafar committed:
[expo-font][android] Rewrite FontLoaderPackage.java to Kotlin

kkafar committed:
[expo-font][android] Rewrite FontLoadedModule.java to Kotlin

kkafar committed:
[expo-font][android] Update changelog

kkafar committed:
Update packages/expo-font/CHANGELOG.md

Co-authored-by: Expo CI <34669131+expo-ci@users.noreply.github.com>

lukmccall commented on packages/expo-font/android/src/main/java/expo/modules/font/FontLoaderModule.kt:
```diff
@@ -0,0 +1,73 @@
+private const val EXPORTED_NAME = "ExpoFontLoader"
+
+class FontLoaderModule(context: Context) : ExportedModule(context) {
+  private var mModuleRegistry: ModuleRegistry? = null
```
Can we use ModuleRegistryDelegate instead?

kkafar committed:
[expo-font][android] Apply spotlessKotlin suggestion

lukmccall commented on packages/expo-font/android/src/main/java/expo/modules/font/FontLoaderModule.kt:
```diff
@@ -0,0 +1,73 @@
+    try {
+      // TODO: remove Expo references
+      // https://github.com/expo/expo/pull/4652#discussion_r296630843
+      var prefix = ""
+      if (isScoped) {
+        prefix = "ExpoFont-"
+      }
```
```suggestion
      val prefix = if (isScoped) {
        prefix = "ExpoFont-"
      } else {
      	""
      }
```

lukmccall reviewed (commented):
LGTM 💪

kkafar committed:
[expo-font][android] Apply suggestion (missing case)

Co-authored-by: Łukasz Kosmaty <kosmatylukasz@gmail.com>

kkafar committed:
[expo-font][android] Fix typos

kkafar committed:
[expo-font][android] Change Kotlin jvmTarget to 1.8 (gradle)

kkafar committed:
[expo-font][android] Replace ModuleRegistry with ModuleRegistryDelegate

These changes are subject to change as they introduce few ugly
null-checks and possible inadequacy in line:

```kotlin
val constantsModule: ConstantsInterface by moduleRegistry()
```

because, concluding from line:

```kotlin
return constantsModule != null && "expo" == constantsModule.appOwnership
```

it is techically possible for `constantsModule` variable to be null ==>
that would require `constantsModule` variable to be declared as below:

```kotlin
val constantsModule: ConstantsInterface? by moduleRegistry()
```

however it is impossible, as such change follows with error in second
part of check:

```kotlin
return constantsModule != null && "expo" == constantsModule.appOwnership
```
where `constantsModule` technically might be null.

kkafar committed:
[expo-font][android] Improve type-nullability in android code

kkafar committed:
[expo-font][android] Apply spotlessKotlin suggestion

lukmccall commented:
@kkafar, it's ready to be merged?

kkafar commented:
I believe so!
